spark3/src/test/java/org/apache/iceberg/spark/source/TestStructuredStreamingRead3.java

```diff
@@ -0,0 +1,435 @@
+          new SimpleRecord(3, "three")),
+      Lists.newArrayList(
+          new SimpleRecord(4, "four"),
+          new SimpleRecord(5, "five")),
+      Lists.newArrayList(
+          new SimpleRecord(6, "six"),
+          new SimpleRecord(7, "seven")));
+
+  /**
+   * test data - to be used for multiple write batches
+   * each batch inturn will have multiple snapshots
+   */
+  private static final List<List<List<SimpleRecord>>> TEST_DATA_MULTIPLE_WRITES_MULTIPLE_SNAPSHOTS = Lists.newArrayList(
+      Lists.newArrayList(
+          Lists.newArrayList(
+              new SimpleRecord(1, "one"),
+              new SimpleRecord(2, "two"),
+              new SimpleRecord(3, "three")),
+          Lists.newArrayList(
+              new SimpleRecord(4, "four"),
+              new SimpleRecord(5, "five"))),
+      Lists.newArrayList(
+          Lists.newArrayList(
+              new SimpleRecord(6, "six"),
+              new SimpleRecord(7, "seven")),
+          Lists.newArrayList(
+              new SimpleRecord(8, "eight"),
+              new SimpleRecord(9, "nine"))),
+      Lists.newArrayList(
+          Lists.newArrayList(
+              new SimpleRecord(10, "ten"),
+              new SimpleRecord(11, "eleven"),
+              new SimpleRecord(12, "twelve")),
+          Lists.newArrayList(
+              new SimpleRecord(13, "thirteen"),
+              new SimpleRecord(14, "fourteen")),
+          Lists.newArrayList(
+              new SimpleRecord(15, "fifteen"),
+              new SimpleRecord(16, "sixteen"))));
+
+  private static SparkSession spark = null;
+
+  @Rule
+  public TemporaryFolder temp = new TemporaryFolder();
+
+  @Rule
+  public ExpectedException exceptionRule = ExpectedException.none();
+
+  @BeforeClass
+  public static void startSpark() {
+    TestStructuredStreamingRead3.spark = SparkSession.builder()
+        .master("local[2]")
+        .config("spark.sql.shuffle.partitions", 4)
+        .getOrCreate();
+  }
+
+  @AfterClass
+  public static void stopSpark() {
+    SparkSession currentSpark = TestStructuredStreamingRead3.spark;
+    TestStructuredStreamingRead3.spark = null;
+    currentSpark.stop();
+  }
+
+  @Before
+  public void setupTable() {
+    sql("CREATE TABLE %s " +
+        "(id INT, data STRING) " +
+        "USING iceberg " +
+        "PARTITIONED BY (bucket(3, id))", tableName);
+    this.table = validationCatalog.loadTable(tableIdent);
+    this.tableIdentifier = tableName;
+  }
+
+  @After
+  public void stopStreams() throws TimeoutException {
+    for (StreamingQuery query : spark.streams().active()) {
+      query.stop();
+    }
+  }
+
+  @After
+  public void removeTables() {
+    sql("DROP TABLE IF EXISTS %s", tableName);
+  }
+
+  @SuppressWarnings("unchecked")
+  @Test
+  public void testReadStreamOnIcebergTableWithMultipleSnapshots() throws Exception {
+    List<List<SimpleRecord>> expected = TEST_DATA_MULTIPLE_SNAPSHOTS;
+    appendDataAsMultipleSnapshots(expected, this.tableIdentifier);
```

Comment:
No need to prefix references with `this.` unless it is an assignment. It helps show that an assignment is updating object state instead of local state, but it usually isn't as valuable when accessing a variable.

Reply:
i am kinda sort of in love with the fact that - how thoroughly rationalized you are.

Reply:
Thanks! I think that a clear rationale is the basis for good reviews. If I can't justify why I'm asking for something, I should let it go. And if I think it's probably better another way then I should always share why I think so.